File: backend/src/main/java/org/springframework/boot/backend/entity/input/RingedBird.java
package org.springframework.boot.backend.entity.input;

import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDate;
import java.time.LocalTime;

/**
 *  Klasa RingedBird treba generirati .psv datoteku opisanu u RA Guideu, za slanje
 *  i primanje podataka
 */

@Entity
@Table(name = "ringed_bird")
@Data
public class RingedBird {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @ManyToOne
    @JoinColumn(name = "ringing_scheme_id")
    private RingingScheme ringingScheme;

    @ManyToOne
    @JoinColumn(name = "primary_identification_method_id")
    private PrimaryIdentificationMethod primaryIdentificationMethod;

    @ManyToOne
    @JoinColumn(name = "ring_code_id")
    private RingCode ringCode; // alphanumeric 10 chars

    @ManyToOne
    private VerificationOfTheMetalRing verificationOfTheMetalRing;
    @ManyToOne
    private MetalRingInformation metalRingInformation;
    @ManyToOne
    private OtherMarksInformation otherMarksInformation;
    @ManyToOne
    private Bird species; // 5 digits
    @ManyToOne
    private Manipulated manipulated;
    @ManyToOne
    private MovedBeforeTheEncounter movedBeforeTheEncounter;
    @ManyToOne
    private CatchingMethods catchingMethods;
    @ManyToOne
    private CatchingLures catchingLures;
    @ManyToOne
    private Sex sex;
    @ManyToOne
    private Age age;
    @ManyToOne
    private Status status;
    @ManyToOne
    private BroodSize broodSize;
    @ManyToOne
    private PullusAge pullusAge;
    @ManyToOne
    private AccuracyOfPullusAge accuracyOfPullusAge;
    private LocalDate date; // 8 digits
    @ManyToOne
    private AccuracyOfDate accuracyOfDate;
    private LocalTime time; // 4 characters
    @ManyToOne
    private Place placeCode; // 4 characters
    private String geoCoordinates; // 15 chars, uzima iz Place
    @ManyToOne
    private AccuracyOfCoordinates accuracyOfCoordinates;
    @ManyToOne
    private Condition condition;
    @ManyToOne
    private Circumstances circumstances;
    @ManyToOne
    private EURINGCodeIdentifier euringCodeIdentifier;
    private String derivedDataDistance; // max 5 digs
    private String derivedDataDirection; // 3 digs with leading zeros
    private String derivedDataElapsedTime; // max 5 digits
    private Double wingLength; // max precision 0.5 mm
    private Double thirdPrimary; // max precision 0.5 mm
    @ManyToOne
    private StateOfWingPoint stateOfWingPoint;
    private Double mass; // max precision 0.1 g
    @ManyToOne
    private Moult moult;
    @ManyToOne
    private PlumageCode plumageCode;
    private Double hindClaw; // maximum precision 0.1 g
    private Double billLength; // maximum precision 0.1 g
    @ManyToOne
    private BillMethod billMethod;
    private Double totalHeadLength; // maximum precision 0.1 g
    private Double tarsus; // maximum precision 0.1 g
    @ManyToOne
    private TarsusMethod tarsusMethod;
    private Double tailLength; // maximum precision 0.1 g
    private Double tailDifference; // maximum precision 0.1 g
    @ManyToOne
    private FatScore fatScore;
    private String fatScoreMethod; // ESF method
    @ManyToOne
    private PectoralMuscleScore pectoralMuscleScore;
    @ManyToOne
    private BroodPatch broodPatch;
    private String primaryScore; // 2 digits; ignore
    @ManyToOne
    private PrimaryMoult primaryMoult; // 10 chars
    private String oldGreaterCoverts;
    @ManyToOne
    private Alula alula;
    @ManyToOne
    private CarpalCovert carpalCovert;

    @ManyToOne
//    @JoinColumn(name = "sexing_method_id")
    private SexingMethod sexingMethod;

    private String placeName; // uzima iz Place objekta
    private String remarks; // opaske
    private String reference; // za referenciranje originalno prstenovane vrste
    private String latitude; //
    private String longitude;
    private String currentPlaceCode; // 4 chars
    private String moreOtherMarks;
}

File: frontend/src/app/interfaces/attributes/ringed-bird.ts
import { Time } from "@angular/common";
import { Bird } from "../../classes/bird.model";
import { AccuracyOfDate } from "./accuracy-of-date";
import { AccuracyOfPullusAge } from "./accuracy-of-pullus-age";
import { Age } from "./age";
import { BroodSize } from "./brood-size";
import { CatchingLures } from "./catching-lures";
import { CatchingMethods } from "./catching-methods";
import { Manipulated } from "./manipulated";
import { MetalRingInformation } from "./metal-ring-information";
import { MovedBeforeTheEncounter } from "./moved-before-the-encounter";
import { OtherMarksInformation } from "./other-marks-information";
import { PrimaryIdentificationMethod } from "./primary-identification-method";
import { RingCode } from "./ring-code";
import { RingingScheme } from "./ringing-scheme";
import { Sex } from "./sex";
import { Status } from "./status";
import { VerificationOfTheMetalRing } from "./verification-of-the-metal-ring";
import { Place } from "./place";
import { Circumstances } from "./circumstances";
import { EuringCodeIdentifier } from "./euring-code-identifier";
import { StateOfWingPoint } from "./state-of-wing-point";
import { Moult } from "./moult";
import { PlumageCode } from "./plumage-code";
import { BillMethod } from "./bill-method";
import { TarsusMethod } from "./tarsus-method";
import { FatScore } from "./fat-score";
import { PectoralMuscleScore } from "./pectoral-muscle-score";
import { BroodPatch } from "./brood-patch";
import { PrimaryMoult } from "./primary-moult";
import { Alula } from "./alula";
import { CarpalCovert } from "./carpal-covert";
import { SexingMethod } from "./sexing-method";
import { PullusAge } from "./pullus-age";
import { Condition } from "./condition";

export interface RingedBird {
    id: number;
    ringingScheme: RingingScheme;
    primaryIdentificationMethod: PrimaryIdentificationMethod;
    ringCode: RingCode;
    verificationOfTheMetalRing: VerificationOfTheMetalRing;
    metalRingInformation: MetalRingInformation;
    otherMarksInformation: OtherMarksInformation;
    species: Bird;
    manipulated: Manipulated;
    movedBeforeTheEncounter: MovedBeforeTheEncounter;
    catchingMethods: CatchingMethods;
    catchingLures: CatchingLures;
    sex: Sex;
    age: Age;
    status: Status;
    broodSize: BroodSize; 
    pullusAge: PullusAge; 
    accuracyOfPullusAge: AccuracyOfPullusAge;
    date: Date;
    accuracyOfDate: AccuracyOfDate;
    time: Time;
    placeCode: Place;
    condition: Condition; 
    circumstances: Circumstances;
    euringCodeIdentifier: EuringCodeIdentifier;
    derivedDataDistance: string;
    derivedDataDirection: string;
    derivedDataElapsedTime: string;
    wingLength: number;
    thirdPrimary: number;
    stateOfWingPoint: StateOfWingPoint;
    mass: number;
    moult: Moult;
    plumageCode: PlumageCode;
    hindClaw: number;
    billLength: number;
    billMethod: BillMethod;
    totalHeadLength: number;
    tarsus: number;
    tarsusMethod: TarsusMethod;
    tailLength: number;
    tailDifference: number;
    fatScore: FatScore;
    fatScoreMethod: string; // ETF 
    pectoralMuscleScore: PectoralMuscleScore;
    broodPatch: BroodPatch;
    primaryMoult: PrimaryMoult;
    oldGreaterCoverts: string;
    alula: Alula;
    carpalCovert: CarpalCovert;
    sexingMethod: SexingMethod;
    remarks: string;
    reference: string;
    moreOtherMarks: string;
  }
  